src/vivarium_census_prl_synth_pop/utilities.py

```diff
@@ -297,89 +267,164 @@ def build_output_dir(output_dir: Path, subdir: Optional[Union[str, Path]] = None
     return output_dir
 
 
+def get_state_puma_map(df: pd.DataFrame) -> Dict[int, int]:
+    return df.groupby("state")["puma"].apply(lambda x: list(x.unique())).to_dict()
+
+
 def update_address_id(
-    df: pd.DataFrame,
-    rows_to_update: pd.Index,
+    units: pd.DataFrame,
+    units_that_move_ids: pd.Index,
     starting_address_id: int,
     address_id_col_name: str = "address_id",
 ) -> pd.DataFrame:
     """
     Parameters
     ----------
-    df: the pd.DataFrame to update, business table
+    units: the pd.DataFrame to update, business table
     rows_to_update: a pd.Index of the rows of df to update
     starting_address_id: int that is tracking value for business or household_ids max value.:
     address_id_col_name: a string. the name of the column in df to hold addresses.
     Returns
     -------
     df with appropriately updated address_ids
     """
-    df.loc[rows_to_update, address_id_col_name] = starting_address_id + np.arange(
-        len(rows_to_update)
+    units.loc[units_that_move_ids, address_id_col_name] = starting_address_id + np.arange(
+        len(units_that_move_ids)
     )
-    return df
+    return units
 
 
-def update_address_id_for_unit_and_sims(
-    pop: pd.DataFrame,
-    moving_units: pd.DataFrame,
+def update_state_and_puma(
+    units: pd.DataFrame,
     units_that_move_ids: pd.Index,
+    state_col_name: str,
+    puma_col_name: str,
+    state_puma_map: Dict[int, List[int]],
+    randomness,
+) -> pd.DataFrame:
+    """Sample from all states/pumas in the artifact. This adds 'state' and 'puma'
+    columns to the units dataframe
+    """
+    state_puma_options = []
+    for state in state_puma_map:
+        state_puma_options.extend([(state, puma) for puma in state_puma_map[state]])
+    state_puma_choices = pd.DataFrame(
+        vectorized_choice(
+            options=np.array(state_puma_options, dtype="i,i"),
+            n_to_choose=len(units_that_move_ids),
+            randomness_stream=randomness,
+            additional_key="sampling_state_puma",
```

Comment:
Since you use this in multiple places, won't it use the same randomness each time? I think you might need to allow passing an `additional_key` to this method.